### PR TITLE
Implement Maxout Network

### LINKED ISSUE #439: Implement Maxout Networks
Reference : Goodfellow, Ian J., et al. "Maxout networks." arXiv preprint arXiv:1302.4389 (2013). [URL](http://arxiv.org/abs/1302.4389)

See also the discussion in Google User Group on Sep. 26th 2015

### PR DESCRIPTION
**Pull request by delta2323**
This PR fixes #439

### CODE REVIEW

File: chainer/links/activation/maxout.py [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+         `URL <http://jmlr.org/proceedings/papers/v28/goodfellow13.html>`_
+    """
+
+    def __init__(self, in_size, num_channel, out_size,
+                 wscale=1, initialW=None, initial_bias=0):
+        super(Maxout, self).__init__(W=(in_size, num_channel, out_size))
+        if initialW is None:
+            initialW = numpy.random.normal(
+                0, wscale * numpy.sqrt(1. / in_size),
+                (in_size, num_channel, out_size))
```
Comment (okuta) [inline chainer/links/activation/maxout.py]: How about use temporary variable?

File: tests/chainer_tests/functions_tests/activation_tests/test_maxout.py [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+            self.y += self.b
+        self.y = numpy.max(self.y, axis=1)
+        self.gy = numpy.random.uniform(
+            -1, 1, self.y.shape).astype(numpy.float32)
+
+    def check_forward(self, x_data, W_data, b_data, y_expect):
+        x = chainer.Variable(x_data)
+        W = chainer.Variable(W_data)
+        if b_data is None:
+            y = functions.maxout(x, W)
```
Comment (okuta) [inline tests/chainer_tests/functions_tests/activation_tests/test_maxout.py]: How about add `b = None`?

File: tests/chainer_tests/functions_tests/activation_tests/test_maxout.py [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+
+        self.check_forward(
+            cuda.to_gpu(self.x), cuda.to_gpu(self.W),
+            b, cuda.to_gpu(self.y))
+
+    def check_backward(self, x_data, W_data, b_data, y_grad):
+        x = chainer.Variable(x_data)
+        W = chainer.Variable(W_data)
+        if b_data is None:
+            y = functions.maxout(x, W)
```
Comment (okuta) [inline tests/chainer_tests/functions_tests/activation_tests/test_maxout.py]: How about add `b = None` for remove `getattr(self, 'b', None)`?

File: chainer/functions/activation/maxout.py [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+    def backward(self, inputs, grad_outputs):
+        gy = grad_outputs[0]
+        x = _as_mat(inputs[0])
+        W = inputs[1]
+
+        xp = cuda.get_array_module(*inputs)
+        # gradient of z = xW + b
+        gz = xp.zeros((gy.shape[0], W.shape[1], gy.shape[1]), x.dtype)
+        gz_r = xp.rollaxis(gz, 1)
+        for idx in numpy.ndindex(gy.shape):
```
Comment (okuta) [inline chainer/functions/activation/maxout.py]: How about below code?
```python
if xp == numpy:
    idx0 = xp.arange(len(gy))[:,None]
    idx1 = xp.arange(gy.shape[1])
    gz[idx0, self.argmax, idx1] = gy
else:
    gz_r = xp.rollaxis(gz, 1)
    cuda.elementwise(
        'T gy, S argmax, int32 n', 'raw T gz',
        'gz[argmax * n + i] = gy', 'maxout_bwd'
    )(gy, self.argmax, gz_r.size // len(gz_r), gz_r)
```

**Comment by okuta** [comment]
I think the unittest is necessary to more stabilize.
Please see TtestMaxPooing2D.
https://github.com/pfnet/chainer/blob/3a7bc835aca856cf08b23ff39ec38c70dbe735e2/tests/chainer_tests/functions_tests/pooling_tests/test_max_pooling_2d.py#L15

File: chainer/functions/activation/maxout.py [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+def maxout(x, W, b=None):
+    """non parametrized Maxout activation function
+
+    It accepts two or three arguments: an input minibatch ``x``,
+    a weight tensor ``W``, and optionally a bias matrix ``b``.
+    It computes
+
+    .. math::
+
+      `Y_{i} = \mathrm{max}_{j} (x^{T}W_{\cdot ij} + b_{ij})`
```
Comment (okuta) [inline chainer/functions/activation/maxout.py]: Please fix backslash. `Y_{i} = \\mathrm{max}_{j} (x^{T}W_{\\cdot ij} + b_{ij})`

File: chainer/functions/activation/maxout.py [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+
+    where :x: is a input vector and :math:`W_{\cdot ij}
```
Comment (okuta) [inline chainer/functions/activation/maxout.py]: Please fix backslash.

File: chainer/functions/activation/maxout.py (hunk 4 above)
Comment (delta2323) [inline chainer/functions/activation/maxout.py]: Thank you for suggestion.

File: tests/chainer_tests/functions_tests/activation_tests/test_maxout.py (hunk 3 above)
Comment (okuta) [inline tests/chainer_tests/functions_tests/activation_tests/test_maxout.py]: Done.

File: tests/chainer_tests/functions_tests/activation_tests/test_maxout.py (hunk 2 above)
Comment (okuta) [inline tests/chainer_tests/functions_tests/activation_tests/test_maxout.py]: Done.

File: chainer/functions/activation/maxout.py [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+    Minibatch dimension is omitted in the above equation.
+
+    Args:
+       x (~chainer.Variable): Input variable. Its first dimension is assumed
+            to be the *minibatch dimension*. The other dimensions are treated
+            as concatenated one dimension whose size must be ``N``.
+       W (~chainer.Variable): Weight variable of shape ``(N, C, M)``
+       b (~chainer.Variable): Bias variable (optional) of shape ``(C, M)```
+    Returns:
+        ~chainer.Variable: Outputvariable
```
Comment (okuta) [inline chainer/functions/activation/maxout.py]: Typo

File: chainer/functions/activation/maxout.py [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+    is a sub-vector extracted from :math:`W` by fixing second
+    and third dimensions to :math:`i` and :math:`j`, respectively.
+    Minibatch dimension is omitted in the above equation.
+
+    Args:
+       x (~chainer.Variable): Input variable. Its first dimension is assumed
+            to be the *minibatch dimension*. The other dimensions are treated
+            as concatenated one dimension whose size must be ``N``.
+       W (~chainer.Variable): Weight variable of shape ``(N, C, M)``
+       b (~chainer.Variable): Bias variable (optional) of shape ``(C, M)```
```
Comment (okuta) [inline chainer/functions/activation/maxout.py]: Please replace back quote to period.

File: chainer/functions/activation/maxout.py [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+    where :math:`x` is a input vector and :math:`W_{\\cdot ij}`
+    is a sub-vector extracted from :math:`W` by fixing second
+    and third dimensions to :math:`i` and :math:`j`, respectively.
+    Minibatch dimension is omitted in the above equation.
+
+    Args:
+       x (~chainer.Variable): Input variable. Its first dimension is assumed
+            to be the *minibatch dimension*. The other dimensions are treated
+            as concatenated one dimension whose size must be ``N``.
+       W (~chainer.Variable): Weight variable of shape ``(N, C, M)``
```
Comment (okuta) [inline chainer/functions/activation/maxout.py]: Please add period.

File: chainer/links/activation/maxout.py [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+        out_size (int): Dimension of output vectors.
+        wscale (float): Scaling factor of the weight matrix.
+        initialW (3-D array): Initial weight value. If ``None``, then this
+            function uses to initialize ``wscale``.
+        initial_bias (2-D array): Initial bias value. If ``None``, then this
+            functions uses to initialize ``bias``.
+
+    Attributes:
+        W (~chainer.Variable): Weight tensor with shape
+            (``in_size``, ``num_channel``, ``out_size``)
```
Comment (okuta) [inline chainer/links/activation/maxout.py]: How about ``` ``(in_size, num_channel, out_size)``.  ```?

File: chainer/links/activation/maxout.py [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+        b (~chainer.Variable): Bias vector with shape
+            (``num_channel``, ``out_size``),
```
Comment (okuta) [inline chainer/links/activation/maxout.py]: How about ``` ``(num_channel, out_size)``.  ```?

**Comment by okuta** [comment]
LGTM except comments.

**Comment by okuta** [comment]
OK!

**Pull request merged**